example_configs/ocis

```diff
@@ -0,0 +1,99 @@
+
+So you must create the bind mount locations and manually chown them to uid/gid 1000, eg.
+
+```
+# cd /srv/ocis
+# mkdir app cfg
+# chown 1000:1000 app cfg
+# docker compose up -d && docker compose logs -f 
+```
+
+## .env
+
+```
+OCIS_URL="https://ocis.example.nz"
+LDAP_BASE_DN="dc=example,dc=nz"
+LDAP_BIND_PASSWORD=very-secret-yogurt
+# LLDAP UUID to be given admin permissions
+LLDAP_ADMIN_UUID=c1c2428a-xxxx-yyyy-zzzz-6cc946bf6809
+```
+
+## docker-compose.yml
+
+```
+version: "3.7"
+
+networks:
+  caddy:
+    external: true
+
+services:
+  ocis:
+    image: owncloud/ocis:5.0.0-rc.4
+    container_name: ocis
+    networks:
+      - caddy
+    entrypoint:
+      - /bin/sh
+    command: ["-c", "ocis init || true; ocis server"]
+    environment:
+      OCIS_URL: ${OCIS_URL}
+      OCIS_LOG_LEVEL: warn
+      OCIS_LOG_COLOR: "false"
+      PROXY_TLS: "false" # do not use SSL between Traefik and oCIS
+      OCIS_INSECURE: "false"
+      # Basic Auth is required for WebDAV clients that don't support OIDC
+      PROXY_ENABLE_BASIC_AUTH: "false"
+      #IDM_ADMIN_PASSWORD: "${ADMIN_PASSWORD}" # Not needed if admin user is in LDAP (?)
+      #OCIS_PASSWORD_POLICY_BANNED_PASSWORDS_LIST: "banned-password-list.txt"
+
+      # Assumes your LLDAP container is named `lldap`
+      OCIS_LDAP_URI: ldap://lldap:3890
+      OCIS_LDAP_INSECURE: "true"
+      OCIS_LDAP_BIND_DN: "uid=admin,ou=people,${LDAP_BASE_DN}"
+      OCIS_LDAP_BIND_PASSWORD: ${LDAP_BIND_PASSWORD}
+      OCIS_ADMIN_USER_ID: ${LLDAP_ADMIN_UUID}
+
+      OCIS_LDAP_USER_ENABLED_ATTRIBUTE: uid
+      GRAPH_LDAP_SERVER_WRITE_ENABLED: "false" # Does your LLDAP bind user have write access?
+      GRAPH_LDAP_REFINT_ENABLED: "false"
+      # Disable the built in LDAP server
+      OCIS_EXCLUDE_RUN_SERVICES: idm
+      # both text and binary cause errors in LLDAP, seems harmless though (?)
+      #IDP_LDAP_UUID_ATTRIBUTE_TYPE: 'text'
+
+      LDAP_LOGIN_ATTRIBUTES: "uid"
+      IDP_LDAP_LOGIN_ATTRIBUTE: "uid"
+      IDP_LDAP_UUID_ATTRIBUTE: "entryuuid"
+      OCIS_LDAP_USER_SCHEMA_ID: "entryuuid"
+      OCIS_LDAP_GROUP_SCHEMA_ID: "uid"
+      OCIS_LDAP_GROUP_SCHEMA_GROUPNAME: "uid"
+
+      OCIS_LDAP_GROUP_BASE_DN: "ou=groups,${LDAP_BASE_DN}"
+      OCIS_LDAP_GROUP_OBJECTCLASS: "groupOfUniqueNames"
+      # can filter which groups are imported, eg: `(&(objectclass=groupOfUniqueNames)(uid=ocis_*))`
+      OCIS_LDAP_GROUP_FILTER: "(objectclass=groupOfUniqueNames)"
+
+      OCIS_LDAP_USER_BASE_DN: "ou=people,${LDAP_BASE_DN}"
+      OCIS_LDAP_USER_OBJECTCLASS: "inetOrgPerson"
+      # Allows all users
+      #OCIS_LDAP_USER_FILTER: "(objectclass=inetOrgPerson)"
+      # Allows users who are in the LLDAP group `ocis_users`
+      OCIS_LDAP_USER_FILTER: "(&(objectclass=person)(memberOf=cn=ocis_users,ou=groups,${LDAP_BASE_DN}))"
+      # NOT WORKING: Used instead of restricting users with OCIS_LDAP_USER_FILTER
```

Comment:
Curious that it's not working, it looks fairly straightforward. Do you have any details?

Reply:
Can't remember, let me try again and remember ...

If I change ...

```
OCIS_LDAP_USER_FILTER: "(objectclass=inetOrgPerson)"
#OCIS_LDAP_USER_FILTER: "(&(objectclass=person)(memberOf=cn=ocis_users,ou=groups,${LDAP_BASE_DN}))"
OCIS_LDAP_DISABLE_USER_MECHANISM: "group"
OCIS_LDAP_DISABLED_USERS_GROUP_DN: "uid=ocis_disabled,ou=groups,${LDAP_BASE_DN}"
```

I get this over and over in the OCIS logs.  I can log into OCIS, but the web app hangs on a splash screen.  The `ocis_disabled` group already exists in LLDAP.

```
ocis  | 2024-02-19T07:00:57.191571167Z {"level":"error","service":"graph","error":"could not initialize graph service","transport":"http","time":"2024-02-19T07:00:57Z","message":"Failed to initialize server"}
ocis  | 2024-02-19T07:00:57.278463335Z {"level":"error","service":"graph","error":"LDAP Result Code 34 \"Invalid DN Syntax\": Unexpected DN format. Got \"uid=ocis_disabled,ou=groups,dc=example,dc=nz\", expected: \"uid=id,ou=people,dc=example,dc=nz\"","local_user_disable":"uid=ocis_disabled,ou=groups,dc=example,dc=nz","time":"2024-02-19T07:00:57Z","message":"error adding group for disabling users"}
```

Not sure what's up with `Unexpected DN format. Got \"uid=ocis_disabled,ou=groups,dc=example,dc=nz\", expected: \"uid=id,ou=people,dc=example,dc=nz\"`

There aren't any useful logs in LLDAP. 

Reading that error again I see it's trying to write to LLDAP.  Changing ...

```
+      GRAPH_LDAP_SERVER_WRITE_ENABLED: "true" # assuming the external ldap is writable
```

Doesn't seem to make a difference.

Reply:
LLDAP would have more useful logs with verbose mode enabled.

Reply:
```
lldap  | 2024-02-19T19:07:27.839235219Z 2024-02-19T19:07:27.251326853+00:00  INFO     LDAP session [ 146ms | 0.15% / 100.00% ]
lldap  | 2024-02-19T19:07:27.839248904Z 2024-02-19T19:07:27.251492998+00:00  INFO     ┝━ LDAP request [ 146ms | 0.10% / 99.79% ]
lldap  | 2024-02-19T19:07:27.839255914Z 2024-02-19T19:07:27.251520215+00:00  DEBUG    │  ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 1, op: BindRequest(LdapBindRequest { dn: "uid=admin,ou=people,dc=example,dc=nz", cred: LdapBindCred::Simple }), ctrl: [] }
lldap  | 2024-02-19T19:07:27.839294785Z 2024-02-19T19:07:27.251526304+00:00  DEBUG    │  ┝━ do_bind [ 146ms | 0.04% / 99.69% ] dn: uid=admin,ou=people,dc=example,dc=nz
lldap  | 2024-02-19T19:07:27.839303723Z 2024-02-19T19:07:27.251547372+00:00  DEBUG    │  │  ┝━ bind [ 145ms | 0.03% / 99.44% ]
lldap  | 2024-02-19T19:07:27.839309994Z 2024-02-19T19:07:27.251556463+00:00  DEBUG    │  │  │  ┝━ get_password_file_for_user [ 239µs | 0.16% ] user_id: UserId("admin")
lldap  | 2024-02-19T19:07:27.839316609Z 2024-02-19T19:07:27.252077940+00:00  DEBUG    │  │  │  ┕━ passwords_match [ 145ms | 99.25% ] username: admin
lldap  | 2024-02-19T19:07:27.839323348Z 2024-02-19T19:07:27.397189311+00:00  DEBUG    │  │  ┝━ get_user_groups [ 301µs | 0.21% ] user_id: "admin"
lldap  | 2024-02-19T19:07:27.839330134Z 2024-02-19T19:07:27.397680128+00:00  DEBUG    │  │  │  ┕━ 🐛 [debug]:  | return: {GroupDetails { group_id: GroupId(1), display_name: "lldap_admin", creation_date: 2023-05-19T09:35:39.078795764, uuid: Uuid("100fcf69-388b-3725-889f-8c496f9cec40"), attributes: [] }, GroupDetails { group_id: GroupId(5), display_name: "calibre", creation_date: 2023-05-20T05:15:14.956410978, uuid: Uuid("59463dc6-5700-3d0d-96bc-432a5aa43f5f"), attributes: [] }, GroupDetails { group_id: GroupId(6), display_name: "ocis_users", creation_date: 2024-02-19T01:08:57.207719922, uuid: Uuid("4afbf888-08c1-336a-8c79-45a7c4570060"), attributes: [] }, GroupDetails { group_id: GroupId(4), display_name: "jellyfin-admin", creation_date: 2023-05-20T03:35:06.456309899, uuid: Uuid("e5c0de8a-25a7-37f1-be06-eb82b97e7d6d"), attributes: [] }}
lldap  | 2024-02-19T19:07:27.839377439Z 2024-02-19T19:07:27.397690259+00:00  DEBUG    │  │  ┕━ 🐛 [debug]: Success!
lldap  | 2024-02-19T19:07:27.839385521Z 2024-02-19T19:07:27.397698008+00:00  DEBUG    │  ┕━ 🐛 [debug]:  | response: BindResponse(LdapBindResponse { res: LdapResult { code: Success, matcheddn: "", message: "", referral: [] }, saslcreds: None })
lldap  | 2024-02-19T19:07:27.839392663Z 2024-02-19T19:07:27.397966112+00:00  INFO     ┕━ LDAP request [ 86.5µs | 0.06% ]
lldap  | 2024-02-19T19:07:27.839399019Z 2024-02-19T19:07:27.397990976+00:00  DEBUG       ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 2, op: AddRequest(LdapAddRequest { dn: "uid=ocis_disabled,ou=groups,dc=example,dc=nz", attributes: [LdapPartialAttribute { atype: "objectClass", vals: ["groupOfNames", "top"] }, LdapPartialAttribute { atype: "member", vals: [""] }] }), ctrl: [] }
lldap  | 2024-02-19T19:07:27.839406575Z 2024-02-19T19:07:27.398005411+00:00  DEBUG       ┕━ 🐛 [debug]:  | response: AddResponse(LdapResult { code: InvalidDNSyntax, matcheddn: "", message: "Unexpected DN format. Got \"uid=ocis_disabled,ou=groups,dc=example,dc=nz\", expected: \"uid=id,ou=people,dc=example,dc=nz\"", referral: [] })



```

Reply:
I see, they're going for "ask for forgiveness, not permission": they're trying to create the group without checking whether it exists (or apparently whether their config allows them to write to the LDAP server).

We have partial support for creating _users_ via LDAP, but not groups, hence the (bad) error message about an invalid DN (it expects a user DN).

It's not crucial for this integration, but we could add support for creating groups (there's an issue for better LDAP support, #518).